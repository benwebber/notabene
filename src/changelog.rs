//! Represent changelogs in Keep a Changelog format.
use serde::{Deserialize, Serialize};

use crate::ir;

/// A changelog in Keep a Changelog format.
#[derive(Debug, Default, Deserialize, Serialize)]
pub struct Changelog {
    pub title: Option<String>,
    pub unreleased: Option<Unreleased>,
    pub releases: Vec<Release>,
}

/// The unreleased section.
#[derive(Debug, Default, Deserialize, Serialize)]
pub struct Unreleased {
    pub url: Option<String>,
    pub changes: Vec<Changes>,
}

/// A release section.
#[derive(Debug, Default, Deserialize, Serialize)]
pub struct Release {
    pub version: String,
    pub url: Option<String>,
    pub date: Option<String>,
    pub yanked: bool,
    pub changes: Vec<Changes>,
}

/// Changes associated with a release.
#[derive(Debug, Deserialize, Serialize)]
pub struct Changes {
    pub kind: String,
    pub changes: Vec<String>,
}

/// Create a `Changelog` from an [`ir::Changelog`].
/// Use the first title and the first unreleased section as `title` and `unreleased`, respectively.
impl From<ir::Changelog> for Changelog {
    fn from(changelog: ir::Changelog) -> Self {
        let mut title: Option<String> = None;
        let mut unreleased: Option<Unreleased> = None;
        let mut releases: Vec<Release> = Vec::new();
        for section in changelog.sections.into_iter() {
            match section {
                ir::Section::Title(t) => {
                    if title.is_none() {
                        title = Some(t.into_inner())
                    }
                }
                ir::Section::Unreleased(u) => {
                    if unreleased.is_none() {
                        unreleased = Some(u.into())
                    }
                }
                ir::Section::Release(r) => releases.push(r.into()),
            }
        }
        Self {
            title,
            unreleased,
            releases,
        }
    }
}

impl From<ir::Unreleased> for Unreleased {
    fn from(unreleased: ir::Unreleased) -> Self {
        Self {
            url: unreleased.url,
            changes: unreleased.changes.into_iter().map(Into::into).collect(),
        }
    }
}

impl From<ir::Release> for Release {
    fn from(release: ir::Release) -> Self {
        Self {
            version: release.version.into_inner(),
            url: release.url,
            date: release.date.map(|d| d.into_inner()),
            yanked: release.yanked.is_some(),
            changes: release.changes.into_iter().map(Into::into).collect(),
        }
    }
}

impl From<ir::Changes> for Changes {
    fn from(changes: ir::Changes) -> Self {
        Self {
            kind: changes.kind.into_inner(),
            changes: changes
                .changes
                .into_iter()
                .map(|c| c.into_inner())
                .collect(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    use crate::span::Span;

    // Spans are not useful here, so create dummy spans.
    macro_rules! spanned {
        ($s:literal) => {
            ir::Spanned::new(Span::new(0, 0), $s.to_string())
        };
    }

    #[test]
    fn test_from_ir() {
        let ir_changelog = ir::Changelog {
            sections: vec![
                ir::Section::Title(spanned!("Title 1")),
                ir::Section::Title(spanned!("Title 2")),
                ir::Section::Title(spanned!("Title 3")),
                ir::Section::Unreleased(ir::Unreleased {
                    url: Some("https://example.org/unreleased/1".to_string()),
                    changes: vec![ir::Changes {
                        kind: spanned!("Added"),
                        heading_span: Span::default(),
                        changes: vec![spanned!("Add foo"), spanned!("Add bar")],
                    }],
                    ..Default::default()
                }),
                ir::Section::Unreleased(ir::Unreleased {
                    url: Some("https://example.org/unreleased/2".to_string()),
                    changes: vec![ir::Changes {
                        kind: spanned!("Added"),
                        heading_span: Span::default(),
                        changes: vec![spanned!("Add baz"), spanned!("Add quux")],
                    }],
                    ..Default::default()
                }),
                ir::Section::Release(ir::Release {
                    version: spanned!("1.0.0"),
                    url: Some("https://example.org/release/1.0.0".to_string()),
                    date: Some(spanned!("2025-01-01")),
                    changes: vec![ir::Changes {
                        kind: spanned!("Changed"),
                        heading_span: Span::default(),
                        changes: vec![spanned!("Change foo"), spanned!("Change bar")],
                    }],
                    ..Default::default()
                }),
                ir::Section::Release(ir::Release {
                    version: spanned!("0.1.0"),
                    url: Some("https://example.org/release/0.1.0".to_string()),
                    date: Some(spanned!("2024-01-01")),
                    yanked: Some(spanned!("[YANKED]")),
                    changes: vec![ir::Changes {
                        kind: spanned!("Changed"),
                        heading_span: Span::default(),
                        changes: vec![spanned!("Change baz"), spanned!("Change quux")],
                    }],
                    ..Default::default()
                }),
            ],
        };
        assert_yaml_snapshot!(Changelog::from(ir_changelog));
    }
}
