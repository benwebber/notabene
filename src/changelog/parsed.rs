//! Borrowed versions of changelog types.
//!
//! These types also contain location information, used by the [`Linter`](crate::linter::Linter).
#[cfg(feature = "serde")]
use serde::Serialize;

use crate::span::{Locator, Span, Spanned};

use super::traits::Release as _;
use super::{owned, traits};

type SpannedStr<'a> = Spanned<&'a str>;

#[derive(Debug, Default, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub struct ParsedChangelog<'a> {
    pub(crate) source: &'a str,
    pub(crate) title: Option<SpannedStr<'a>>,
    pub(crate) unreleased: Option<ParsedUnreleased<'a>>,
    pub(crate) releases: Vec<ParsedRelease<'a>>,
    pub(crate) invalid_spans: Vec<InvalidSpan>,
}

#[derive(Debug, Default, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub struct ParsedUnreleased<'a> {
    pub(crate) heading_span: Span,
    pub(crate) url: Option<String>,
    pub(crate) changes: Vec<ParsedChanges<'a>>,
}

#[derive(Debug, Default, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub struct ParsedRelease<'a> {
    pub(crate) heading_span: Span,
    pub(crate) version: SpannedStr<'a>,
    pub(crate) url: Option<String>,
    pub(crate) date: Option<SpannedStr<'a>>,
    pub(crate) yanked: Option<SpannedStr<'a>>,
    pub(crate) changes: Vec<ParsedChanges<'a>>,
}

#[derive(Debug, Default, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub struct ParsedChanges<'a> {
    pub(crate) heading_span: Span,
    pub(crate) kind: SpannedStr<'a>,
    pub(crate) items: Vec<SpannedStr<'a>>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub enum InvalidSpan {
    InvalidTitle(Span),
    InvalidSectionHeading(Span),
    UndefinedLinkReference(Span),
    DuplicateUnreleased(Span),
    DuplicateTitle(Span),
    InvalidUnreleasedPosition(Span),
}

impl<'a> traits::Changelog for ParsedChangelog<'a> {
    type Unreleased = ParsedUnreleased<'a>;
    type Release = ParsedRelease<'a>;

    fn title(&self) -> Option<&str> {
        self.title.as_deref()
    }

    fn unreleased(&self) -> Option<&Self::Unreleased> {
        self.unreleased.as_ref()
    }

    fn releases(&self) -> &[Self::Release] {
        &self.releases
    }
}

impl<'a> traits::Unreleased for ParsedUnreleased<'a> {
    type Changes = ParsedChanges<'a>;

    fn url(&self) -> Option<&str> {
        self.url.as_deref()
    }

    fn changes(&self) -> &[Self::Changes] {
        &self.changes
    }
}

impl<'a> traits::Release for ParsedRelease<'a> {
    type Changes = ParsedChanges<'a>;

    fn version(&self) -> &str {
        &self.version
    }

    fn url(&self) -> Option<&str> {
        self.url.as_deref()
    }

    fn date(&self) -> Option<&str> {
        self.date.as_deref()
    }

    fn yanked(&self) -> bool {
        self.yanked.is_some_and(|s| s.value == "[YANKED]")
    }

    fn changes(&self) -> &[Self::Changes] {
        &self.changes
    }
}

impl<'a> traits::Changes for ParsedChanges<'a> {
    fn kind(&self) -> &str {
        self.kind.value
    }

    fn items(&self) -> impl Iterator<Item = &str> {
        self.items.iter().map(|i| &**i)
    }
}

impl<'a> ParsedChangelog<'a> {
    pub fn parse(s: &'a str) -> Self {
        crate::parser::parse(s)
    }

    pub(crate) fn locator(&self) -> Locator<'a> {
        Locator::new(self.source)
    }

    pub(crate) fn locate(
        &self,
        diagnostic: &crate::Diagnostic<Span>,
    ) -> crate::Diagnostic<crate::span::Position> {
        self.locator().locate(diagnostic)
    }

    pub fn locate_all(
        &self,
        diagnostics: &[crate::Diagnostic],
    ) -> Vec<crate::Diagnostic<crate::span::Position>> {
        diagnostics.iter().map(|d| self.locate(d)).collect()
    }

    pub fn to_owned(&self) -> owned::OwnedChangelog {
        owned::OwnedChangelog {
            title: self.title.map(|s| s.value.to_owned()),
            unreleased: self.unreleased.as_ref().map(|u| u.to_owned()),
            releases: self.releases.iter().map(|r| r.to_owned()).collect(),
        }
    }
}

impl<'a> ParsedUnreleased<'a> {
    pub fn to_owned(&self) -> owned::OwnedUnreleased {
        owned::OwnedUnreleased {
            url: self.url.clone(),
            changes: self.changes.iter().map(|c| c.to_owned()).collect(),
        }
    }
}

impl<'a> ParsedRelease<'a> {
    pub fn to_owned(&self) -> owned::OwnedRelease {
        owned::OwnedRelease {
            version: self.version.value.to_owned(),
            url: self.url.clone(),
            date: self.date.map(|s| s.value.to_owned()),
            yanked: self.yanked(),
            changes: self.changes.iter().map(|c| c.to_owned()).collect(),
        }
    }
}

impl<'a> ParsedChanges<'a> {
    pub fn to_owned(&self) -> owned::OwnedChanges {
        owned::OwnedChanges {
            kind: self.kind.value.to_owned(),
            items: self.items.iter().map(|i| i.value.to_owned()).collect(),
        }
    }
}
