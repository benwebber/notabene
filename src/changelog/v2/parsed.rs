use crate::ir;
use crate::span::{Span, Spanned};

use super::traits::Release as _;
use super::{owned, traits};

type SpannedStr<'a> = Spanned<&'a str>;

pub struct Changelog<'a> {
    pub(crate) title: Option<SpannedStr<'a>>,
    pub(crate) unreleased: Option<Unreleased<'a>>,
    pub(crate) releases: Vec<Release<'a>>,
    pub(crate) invalid_spans: Vec<InvalidSpan>,
}

pub struct Unreleased<'a> {
    pub(crate) heading_span: Span,
    pub(crate) url: Option<SpannedStr<'a>>,
    pub(crate) changes: Vec<Changes<'a>>,
}

pub struct Release<'a> {
    pub(crate) heading_span: Span,
    pub(crate) version: SpannedStr<'a>,
    pub(crate) url: Option<SpannedStr<'a>>,
    pub(crate) date: Option<SpannedStr<'a>>,
    pub(crate) yanked: Option<SpannedStr<'a>>,
    pub(crate) changes: Vec<Changes<'a>>,
}

pub struct Changes<'a> {
    pub(crate) heading_span: Span,
    pub(crate) kind: SpannedStr<'a>,
    pub(crate) items: Vec<SpannedStr<'a>>,
}

pub enum InvalidSpan {
    InvalidTitle(Span),
    InvalidHeading(Span),
    InvalidLinkReference(Span),
    DuplicateUnreleased(Span),
}

impl<'a> traits::Changelog for Changelog<'a> {
    type Unreleased = Unreleased<'a>;
    type Release = Release<'a>;

    fn title(&self) -> Option<&str> {
        self.title.as_deref()
    }

    fn unreleased(&self) -> Option<&Self::Unreleased> {
        self.unreleased.as_ref()
    }

    fn releases(&self) -> &[Self::Release] {
        &self.releases
    }
}

impl<'a> traits::Unreleased for Unreleased<'a> {
    type Changes = Changes<'a>;

    fn url(&self) -> Option<&str> {
        self.url.as_deref()
    }

    fn changes(&self) -> &[Self::Changes] {
        &self.changes
    }
}

impl<'a> traits::Release for Release<'a> {
    type Changes = Changes<'a>;

    fn version(&self) -> &str {
        &self.version
    }

    fn url(&self) -> Option<&str> {
        self.url.as_deref()
    }

    fn date(&self) -> Option<&str> {
        self.date.as_deref()
    }

    fn yanked(&self) -> bool {
        self.yanked.map_or(false, |s| s.value == "[YANKED]")
    }

    fn changes(&self) -> &[Self::Changes] {
        &self.changes
    }
}

impl<'a> traits::Changes for Changes<'a> {
    fn kind(&self) -> &str {
        &self.kind.value
    }

    fn items(&self) -> impl Iterator<Item = &str> {
        self.items.iter().map(|i| &**i)
    }
}

impl<'a> Changelog<'a> {
    fn to_owned(&self) -> owned::Changelog {
        owned::Changelog {
            title: self.title.map(|s| s.value.to_owned()),
            unreleased: self.unreleased.as_ref().map(|u| u.to_owned()),
            releases: self.releases.iter().map(|r| r.to_owned()).collect(),
        }
    }
}

impl<'a> Unreleased<'a> {
    fn to_owned(&self) -> owned::Unreleased {
        owned::Unreleased {
            url: self.url.map(|s| s.value.to_owned()),
            changes: self.changes.iter().map(|c| c.to_owned()).collect(),
        }
    }
}

impl<'a> Release<'a> {
    fn to_owned(&self) -> owned::Release {
        owned::Release {
            version: self.version.value.to_owned(),
            url: self.url.map(|s| s.value.to_owned()),
            date: self.date.map(|s| s.value.to_owned()),
            yanked: self.yanked(),
            changes: self.changes.iter().map(|c| c.to_owned()).collect(),
        }
    }
}

impl<'a> Changes<'a> {
    fn to_owned(&self) -> owned::Changes {
        owned::Changes {
            kind: self.kind.value.to_owned(),
            items: self.items.iter().map(|i| i.value.to_owned()).collect(),
        }
    }
}

impl<'a> From<&'a ir::Changelog<'a>> for Changelog<'a> {
    fn from(ir: &'a ir::Changelog<'a>) -> Self {
        let mut title: Option<SpannedStr<'a>> = None;
        let mut unreleased: Option<Unreleased<'a>> = None;
        let mut releases: Vec<Release<'a>> = Vec::new();
        let mut invalid_spans: Vec<InvalidSpan> = Vec::new();
        for section in ir.sections.iter() {
            match section {
                ir::Section::Title(t) => {
                    if title.is_none() {
                        title = Some(SpannedStr {
                            span: t.span,
                            value: t.value,
                        })
                    }
                }
                ir::Section::Unreleased(u) => match unreleased {
                    Some(_) => invalid_spans.push(InvalidSpan::DuplicateUnreleased(u.heading_span)),
                    None => unreleased = Some(u.into()),
                },
                ir::Section::Release(r) => releases.push(r.into()),
                ir::Section::Invalid(i) => {
                    invalid_spans.push(InvalidSpan::InvalidHeading(i.heading_span))
                }
                ir::Section::InvalidTitle(i) => {
                    invalid_spans.push(InvalidSpan::InvalidTitle(i.heading_span))
                }
            }
        }
        let mut broken_links = ir
            .broken_links
            .iter()
            .map(|s| InvalidSpan::InvalidLinkReference(*s))
            .collect();
        invalid_spans.append(&mut broken_links);
        Self {
            title,
            unreleased,
            releases,
            invalid_spans,
        }
    }
}

impl<'a> From<&'a ir::Unreleased<'a>> for Unreleased<'a> {
    fn from(ir: &'a ir::Unreleased<'a>) -> Self {
        // TODO: Get proper span for URL.
        let url = match &ir.url {
            Some(s) => Some(SpannedStr {
                span: Span::default(),
                value: s.as_str(),
            }),
            None => None,
        };
        let changes = ir.changes.iter().map(|c| Changes::from(c)).collect();
        Self {
            heading_span: ir.heading_span,
            url,
            changes,
        }
    }
}

impl<'a> From<&'a ir::Release<'a>> for Release<'a> {
    fn from(ir: &'a ir::Release<'a>) -> Self {
        let version = SpannedStr {
            span: ir.version.span,
            value: ir.version.value,
        };
        // TODO: Get proper span for URL.
        let url = match &ir.url {
            Some(s) => Some(SpannedStr {
                span: Span::default(),
                value: s.as_str(),
            }),
            None => None,
        };
        let date = match &ir.date {
            Some(s) => Some(SpannedStr {
                span: s.span,
                value: s.value,
            }),
            None => None,
        };
        let yanked = match &ir.yanked {
            Some(s) => Some(SpannedStr {
                span: s.span,
                value: s.value,
            }),
            None => None,
        };
        let changes = ir.changes.iter().map(|c| Changes::from(c)).collect();
        Self {
            heading_span: ir.heading_span,
            version,
            url,
            date,
            yanked,
            changes,
        }
    }
}

impl<'a> From<&'a ir::Changes<'a>> for Changes<'a> {
    fn from(ir: &'a ir::Changes<'a>) -> Self {
        let kind = SpannedStr {
            span: ir.kind.span,
            value: ir.kind.value,
        };
        let items = ir
            .changes
            .iter()
            .map(|s| SpannedStr {
                span: s.span,
                value: s.value,
            })
            .collect();
        Self {
            heading_span: ir.heading_span,
            kind,
            items,
        }
    }
}
