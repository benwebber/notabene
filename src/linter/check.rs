use crate::diagnostic::Diagnostic;
use crate::ir;
use crate::location::Location;
use crate::rule::Rule;
use crate::span::Span;

pub(crate) trait Check {
    /// Return the `Rule` this check evaluates.
    fn rule(&self) -> Rule;

    /// Return spans that this check matched.
    fn spans(&self) -> &[Span] {
        &[]
    }

    /// Evaluate this check for the entire `Changelog`.
    fn visit_changelog(&mut self, _changelog: &ir::Changelog) {}

    /// Evaluate this check for a `Section`.
    fn visit_section(&mut self, _section: &ir::Section) {}

    /// Evaluate this check for a `Changes`.
    fn visit_changes(&mut self, _changes: &ir::Changes) {}

    /// Publish diagnostics generated by this check.
    fn diagnostics(&self) -> Vec<Diagnostic> {
        self.spans()
            .iter()
            .map(|span| Diagnostic::new(self.rule(), Some(Location::Span(*span))))
            .collect()
    }
}
