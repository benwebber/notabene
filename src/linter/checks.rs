mod content;
mod links;
pub(crate) mod preamble;
mod releases;
mod structure;

pub use content::*;
pub use links::*;
pub use releases::*;
pub use structure::*;
