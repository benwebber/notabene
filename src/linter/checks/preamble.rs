pub(crate) use crate::linter::Context;
pub(crate) use crate::linter::check::Check;
pub(crate) use crate::rule::Rule;
pub(crate) use crate::span::Span;

pub(crate) use crate::changelog::parsed;
