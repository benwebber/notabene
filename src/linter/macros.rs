macro_rules! invalid_span {
    ($check:ident,$rule:path,$span:path) => {
        #[derive(Default)]
        pub struct $check {
            spans: Vec<Span>,
        }

        impl Check for $check {
            fn rule(&self) -> Rule {
                $rule
            }

            fn spans(&self) -> &[Span] {
                self.spans.as_slice()
            }

            fn visit_invalid_span(&mut self, span: &parsed::InvalidSpan) {
                if let $span(s) = span {
                    self.spans.push(*s);
                }
            }
        }
    };
}
