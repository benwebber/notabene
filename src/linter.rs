//! Linter implementation.
use std::path::PathBuf;

use crate::changelog::parsed;
use crate::changelog::traits::*;
use crate::diagnostic::Diagnostic;
use crate::ruleset::RuleSet;

mod check;
mod checks;

use check::Check;

/// A changelog linter.
#[derive(Debug)]
pub struct Linter<'a> {
    ruleset: &'a RuleSet,
    filename: Option<PathBuf>,
}

impl<'a> Linter<'a> {
    /// Create a new linter with the given ruleset.
    pub fn new(ruleset: &'a RuleSet) -> Self {
        Self {
            ruleset,
            filename: None,
        }
    }

    /// Set the filename reported in diagnostics.
    pub fn with_filename<P: Into<PathBuf>>(self, filename: Option<P>) -> Self {
        Self {
            filename: filename.map(|n| n.into()),
            ..self
        }
    }

    /// Lint a changelog.
    pub fn lint(&self, changelog: &parsed::Changelog) -> Vec<Diagnostic> {
        let mut diagnostics: Vec<Diagnostic> = Vec::new();
        let mut checks: Vec<_> = checks()
            .into_iter()
            .filter(|check| self.ruleset.is_enabled(check.rule()))
            .collect();
        for check in checks.iter_mut() {
            check.visit_changelog(&changelog);
            if let Some(unreleased) = changelog.unreleased() {
                check.visit_unreleased(&unreleased);
                for changes in unreleased.changes() {
                    check.visit_changes(changes);
                }
            }
            for span in &changelog.invalid_spans {
                check.visit_invalid_span(span);
            }
        }
        for release in changelog.releases() {
            for check in checks.iter_mut() {
                check.visit_release(release);
                for changes in release.changes() {
                    check.visit_changes(changes);
                }
            }
        }
        for check in checks.iter_mut() {
            diagnostics.append(
                &mut check
                    .diagnostics()
                    .into_iter()
                    .map(|mut d| {
                        d.path = self.filename.clone();
                        d
                    })
                    .collect(),
            );
        }
        diagnostics
    }
}

/// Lint a changelog with the default ruleset.
pub fn lint(changelog: &parsed::Changelog) -> Vec<Diagnostic> {
    let ruleset = RuleSet::default();
    let linter = Linter::new(&ruleset);
    linter.lint(changelog)
}

fn checks() -> Vec<Box<dyn Check>> {
    vec![
        // E000 Document
        Box::new(checks::MissingTitle::default()),
        Box::new(checks::InvalidTitle::default()),
        Box::new(checks::DuplicateTitle::default()),
        Box::new(checks::InvalidSectionHeading::default()),
        Box::new(checks::UnreleasedOutOfOrder::default()),
        // E100 Unreleased
        Box::new(checks::MissingUnreleased::default()),
        Box::new(checks::DuplicateUnreleased::default()),
        // E200 Release
        Box::new(checks::InvalidDate::default()),
        Box::new(checks::InvalidYanked::default()),
        Box::new(checks::MissingDate::default()),
        Box::new(checks::ReleaseOutOfOrder::default()),
        Box::new(checks::DuplicateVersion::default()),
        // E300 Changes
        Box::new(checks::InvalidChangeType::default()),
        Box::new(checks::DuplicateChangeType::default()),
        // E400 Content
        Box::new(checks::EmptySection::default()),
        // E500 Links
        Box::new(checks::LinkReferenceDoesNotExist::default()),
    ]
}
