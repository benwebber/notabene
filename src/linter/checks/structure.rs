//! `E000` Structure
use super::preamble::*;

#[derive(Default)]
pub struct MissingTitle {
    found: bool,
}

impl Check for MissingTitle {
    fn rule(&self) -> Rule {
        Rule::MissingTitle
    }

    fn visit_changelog(&mut self, changelog: &parsed::ParsedChangelog) {
        if self.found {
            return;
        }
        if changelog.title().is_some() {
            self.found = true;
        }
    }

    fn diagnostics(&self) -> Vec<Diagnostic> {
        if self.found {
            vec![]
        } else {
            vec![Diagnostic::new(self.rule(), None)]
        }
    }
}

invalid_span!(
    DuplicateTitle,
    Rule::DuplicateTitle,
    parsed::InvalidSpan::DuplicateTitle
);

#[derive(Default)]
pub struct MissingUnreleased {
    found: bool,
}

impl Check for MissingUnreleased {
    fn rule(&self) -> Rule {
        Rule::MissingUnreleased
    }

    fn visit_changelog(&mut self, changelog: &parsed::ParsedChangelog) {
        self.found = changelog.unreleased.is_some();
    }

    fn diagnostics(&self) -> Vec<Diagnostic> {
        if self.found {
            vec![]
        } else {
            vec![Diagnostic::new(self.rule(), None)]
        }
    }
}

invalid_span!(
    DuplicateUnreleased,
    Rule::DuplicateUnreleased,
    parsed::InvalidSpan::DuplicateUnreleased
);

invalid_span!(
    UnreleasedOutOfOrder,
    Rule::InvalidUnreleasedPosition,
    parsed::InvalidSpan::UnreleasedOutOfOrder
);

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    use crate::changelog::parsed::{InvalidSpan, ParsedChangelog, ParsedUnreleased};
    use crate::linter::Linter;
    use crate::ruleset::RuleSet;
    use crate::span::{Span, Spanned};

    #[test]
    fn test_missing_title() {
        let ruleset = RuleSet::from([Rule::MissingTitle]);
        let linter = Linter::new(&ruleset);

        let changelog = ParsedChangelog::default();
        assert_yaml_snapshot!(linter.lint(&changelog));

        let changelog = ParsedChangelog {
            title: Some(Spanned::<&str>::default()),
            ..Default::default()
        };
        assert_yaml_snapshot!(linter.lint(&changelog));
    }

    #[test]
    fn test_missing_unreleased() {
        let ruleset = RuleSet::from([Rule::MissingUnreleased]);
        let linter = Linter::new(&ruleset);

        let changelog = ParsedChangelog::default();
        assert_yaml_snapshot!(linter.lint(&changelog));

        let changelog = ParsedChangelog {
            unreleased: Some(ParsedUnreleased::default()),
            ..Default::default()
        };
        assert_yaml_snapshot!(linter.lint(&changelog));
    }

    #[test]
    fn test_duplicate_unreleased() {
        let ruleset = RuleSet::from([Rule::DuplicateUnreleased]);
        let linter = Linter::new(&ruleset);

        let changelog = ParsedChangelog::default();
        assert_yaml_snapshot!(linter.lint(&changelog));

        let changelog = ParsedChangelog {
            invalid_spans: vec![InvalidSpan::DuplicateUnreleased(Span::new(1, usize::MAX))],
            ..Default::default()
        };
        assert_yaml_snapshot!(linter.lint(&changelog));
    }

    #[test]
    fn test_duplicate_title() {
        let ruleset = RuleSet::from([Rule::DuplicateTitle]);
        let linter = Linter::new(&ruleset);

        let changelog = ParsedChangelog::default();
        assert_yaml_snapshot!(linter.lint(&changelog));

        let changelog = ParsedChangelog {
            invalid_spans: vec![InvalidSpan::DuplicateTitle(Span::new(1, usize::MAX))],
            ..Default::default()
        };
        assert_yaml_snapshot!(linter.lint(&changelog));
    }

    #[test]
    fn test_unreleased_not_first() {
        let ruleset = RuleSet::from([Rule::InvalidUnreleasedPosition]);
        let linter = Linter::new(&ruleset);

        let changelog = ParsedChangelog::default();
        assert_yaml_snapshot!(linter.lint(&changelog));

        let changelog = ParsedChangelog {
            invalid_spans: vec![InvalidSpan::UnreleasedOutOfOrder(Span::new(1, usize::MAX))],
            ..Default::default()
        };
        assert_yaml_snapshot!(linter.lint(&changelog));
    }
}
