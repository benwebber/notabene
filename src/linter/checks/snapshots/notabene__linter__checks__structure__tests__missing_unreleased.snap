---
source: src/linter/checks/structure.rs
expression: linter.lint(&changelog)
---
- rule: MissingUnreleased
  location: ~
  path: ~
